extensions/notebook-renderers/src/index.ts

```diff
@@ -203,41 +200,69 @@ function renderError(
 	return disposableStore;
 }
 
-function renderStream(outputInfo: OutputItem, container: HTMLElement, error: boolean, ctx: IRichRenderContext): IDisposable {
-	const disposableStore = createDisposableStore();
-
+function getPreviousOutputWithMatchingMimeType(container: HTMLElement, mimeType: string) {
 	const outputContainer = container.parentElement;
-	if (!outputContainer) {
-		// should never happen
-		return disposableStore;
+
+	const previous = outputContainer?.previousSibling;
+	if (previous) {
+		const outputElement = (previous.firstChild as HTMLElement | null);
+		if (outputElement && outputElement.getAttribute('output-mime-type') === mimeType) {
+			return outputElement;
+		}
 	}
+	return undefined;
+}
+
+// if there is a scrollable output, it will be scrolled to the given value if provided or the bottom of the element
+function appendChildAndScroll(container: HTMLElement, child: HTMLElement, scrollTop?: number) {
+	container.appendChild(child);
+	child.childNodes.forEach((node) => {
+		if (node instanceof HTMLElement && node.classList.contains(scrollableClass)) {
+			node.scrollTop = scrollTop !== undefined ? scrollTop : node.scrollHeight;
+		}
+	});
+}
 
-	const prev = outputContainer.previousSibling;
-	if (prev) {
-		// OutputItem in the same cell
-		// check if the previous item is a stream
-		const outputElement = (prev.firstChild as HTMLElement | null);
-		if (outputElement && outputElement.getAttribute('output-mime-type') === outputInfo.mime) {
-			// same stream
-
-			// find child with same id
-			const existing = outputElement.querySelector(`[output-item-id="${outputInfo.id}"]`) as HTMLElement | null;
-			if (existing) {
-				clearContainer(existing);
+// Find the scrollTop of the existing scrollable output, return undefined if at the bottom
+function findScrolledHeight(outputContainer: HTMLElement): number | undefined {
+	let result: number | undefined;
+	outputContainer.childNodes.forEach((output) => {
```

Comment:
Based on our discussion and testing offline, the DOM structure of streaming output is always

```
div.output-container
  span.output-stream
    <optional> span "Output exceeds the maximium size limit"
    div.scrollable
```

we should be able to do lookup through `querySelector` or checking the lastChild of `span.output-stream` other than nested for loops.